File: androidApp/src/main/java/com/medhelp/medhelp/bg/service/MyFirebaseMessagingService.java
package com.medhelp.medhelp.bg.service;

import android.app.NotificationManager;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.medhelp.medhelp.bg.notifiaction.SimpleNotificationForFCM;
import com.medhelp.medhelp.data.model.chat.SimpleResBoolean;
import com.medhelp.medhelp.data.network.NetworkManager;
import com.medhelp.medhelp.data.pref.PreferencesManager;
import com.medhelp.medhelp.utils.timber_log.LoggingTree;
import com.medhelp.shared.model.UserResponse;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class MyFirebaseMessagingService extends FirebaseMessagingService {

    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (remoteMessage.getNotification() != null) {
                    //если открыта страница то она обновиться без уведомлений
                    if (remoteMessage.getData().size() > 0){
                        String type_message = remoteMessage.getData().get("type_message");
                        String id_kl = remoteMessage.getData().get("id_kl");
                        String id_filial = remoteMessage.getData().get("id_filial");

                        if(type_message!=null && id_kl!=null && id_filial!=null){
                            SimpleNotificationForFCM simpleNotificationForFCM=new SimpleNotificationForFCM(getApplicationContext(), (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE));
                            simpleNotificationForFCM.showData(remoteMessage.getNotification().getTitle(),remoteMessage.getNotification().getBody(),type_message,id_kl,id_filial);
                        }
                    }
                }
            }
        });
    }

    @Override
    public void onNewToken(String token) {
        // If you want to send messages to this application instance or
        // manage this apps subscriptions on the server side, send the
        // Instance ID token to your app server.
        sendRegistrationToServer(token);
    }

    private void sendRegistrationToServer(String token) {
        PreferencesManager prefManager = new PreferencesManager(getApplicationContext());
        NetworkManager networkManager = new NetworkManager(prefManager);

        List<UserResponse> list = prefManager.getUsersLogin();
        if(list==null || list.size()==0){
            return;
        }

        Observable.fromIterable(list)
                .flatMap(userResponse ->
                        networkManager.sendFcmId(String.valueOf(userResponse.getIdUser()),String.valueOf(userResponse.getIdBranch()),token))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<SimpleResBoolean>() {
                               @Override
                               public void onSubscribe(Disposable d) {}

                               @Override
                               public void onNext(SimpleResBoolean userResponse) {}

                               @Override
                               public void onError(Throwable e) {
                                   Timber.tag("my").e(LoggingTree.getMessageForError(e, "MyFirebaseMessagingService$sendFcmToken"));
                               }

                               @Override
                               public void onComplete() {}
                           }
                );
    }

}

File: androidApp/src/main/java/com/medhelp/medhelp/bg/service/MyFirebaseMessagingService.kt
package com.medhelp.medhelp.bg.service

import com.medhelp.medhelp.utils.timber_log.LoggingTree.Companion.getMessageForError
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import android.os.Looper
import com.medhelp.medhelp.bg.notifiaction.SimpleNotificationForFCM
import android.app.NotificationManager
import android.os.Handler
import com.medhelp.medhelp.data.pref.PreferencesManager
import timber.log.Timber
import com.medhelp.shared.network.NetworkManager
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class MyFirebaseMessagingService : FirebaseMessagingService() {
    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        val handler = Handler(Looper.getMainLooper())
        handler.post {
            if (remoteMessage.notification != null) {
                //если открыта страница то она обновиться без уведомлений
                if (remoteMessage.data.size > 0) {
                    val type_message = remoteMessage.data["type_message"]
                    val id_kl = remoteMessage.data["id_kl"]
                    val id_filial = remoteMessage.data["id_filial"]
                    if (type_message != null && id_kl != null && id_filial != null) {
                        val simpleNotificationForFCM = SimpleNotificationForFCM(
                            applicationContext, getSystemService(
                                NOTIFICATION_SERVICE
                            ) as NotificationManager
                        )
                        simpleNotificationForFCM.showData(
                            remoteMessage.notification!!.title, remoteMessage.notification!!
                                .body, type_message, id_kl, id_filial
                        )
                    }
                }
            }
        }
    }

    override fun onNewToken(token: String) {
        // If you want to send messages to this application instance or
        // manage this apps subscriptions on the server side, send the
        // Instance ID token to your app server.
        sendRegistrationToServer(token)
    }


    private var countFcmSend=0
    private fun sendRegistrationToServer(token: String) {
        val prefManager = PreferencesManager(applicationContext)
        val networkManager = NetworkManager()
        val list = prefManager.usersLogin
        if (list == null || list.size == 0) {
            return
        }

        val mainScope = MainScope()

        for (i in list) {
            mainScope.launch {
                kotlin.runCatching {
                    networkManager.sendFcmId(i.idUser.toString(), i.idBranch.toString(), token, prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
                }
                    .onSuccess {
                        mainScope.cancel()
                    }.onFailure {
                        Timber.tag("my").e(getMessageForError(it, "MyFirebaseMessagingService\$sendFcmToken"))
                        mainScope.cancel()
                    }
            }
        }
    }

}